src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt

```diff
@@ -0,0 +1,206 @@
+package at.hannibal2.skyhanni.features.inventory
+
+import at.hannibal2.skyhanni.skyhannimodule.SkyHanniModule
+import at.hannibal2.skyhanni.SkyHanniMod
+import at.hannibal2.skyhanni.events.InventoryUpdatedEvent
+import at.hannibal2.skyhanni.events.LorenzToolTipEvent
+import at.hannibal2.skyhanni.utils.InventoryUtils
+import at.hannibal2.skyhanni.utils.NumberUtil.roundTo
+import at.hannibal2.skyhanni.utils.RegexUtils.matches
+import at.hannibal2.skyhanni.utils.StringUtils.removeColor
+import at.hannibal2.skyhanni.utils.repopatterns.RepoPattern
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+
+@SkyHanniModule
+object FannCost {
+
+    private val config get() = SkyHanniMod.feature.inventory.fannCost
+    private val showCoins get() = config.coinsPerXP
+    private val showBits get() = config.xpPerBit
+    private var trainingMode: TrainingMode = TrainingMode.DAY_COUNT
+
+    private val patternGroup = RepoPattern.group("fann.inventory")
+    private val trainingSlotInventoryPattern by patternGroup.pattern(
+        "training",
+        "Training Slot [1-3]",
+    )
+    private val anvilPattern by patternGroup.pattern(
+        "anvil",
+        "§aBegin Training",
+    )
+    private val expEarnedPattern by patternGroup.pattern(
+        "exp.total",
+        """Will earn a total of (\d{1,3}(?:,\d{3})*(?:\.\d+)?|\d+\.?\d*) EXP\.?""",
+    )
+    private val dailyExpPattern by patternGroup.pattern(
+        "exp.daily",
+        """EXP Per Day: ([\d,]+)(?: \(\+\d{1,2}(\.\d{1,2})?%\))?""",
+    )
+    private val durationPattern by patternGroup.pattern(
+        "training.duration.pattern",
+        """Will take: (?<day>\d+)d (?<hr>\d{1,2})h (?<min>\d{1,2})m (?<sec>\d{1,2})s""",
+    )
+    private val coinsPattern by patternGroup.pattern(
+        "coin",
+        """(\d{1,3}(?:,\d{3})*(?:\.\d+)?|\d+\.?\d*) Coins(?: \([1-5]% off\))?""",
+    )
+    private val bitsPattern by patternGroup.pattern(
+        "bits",
+        """(\d{1,3}(?:,\d{3})*(?:\.\d+)?|\d+\.?\d*) Bits""",
+    )
+    private val desiredLevelPatter by patternGroup.pattern(
+        "slot24.name.level",
+        "Desired Level: (200|1?[0-9]?[0-9])",
+    )
+    private val userInputPattern by patternGroup.pattern(
+        "slot24.name.input",
+        "User Input",
+    )
+    private val trainingTypePattern by patternGroup.pattern(
+        "training.type",
+        "Type: (Free|Light|Moderate|Expert|Ultra|Turbo!)",
+    )
```

Comment:
all of these need regex tests, they also need named pattern groups and none of them should include triple quotes